File: tcb_manufacturing_customizations/public/js/stock_reconciliation.js
frappe.ui.form.on("Stock Reconciliation",{
    custom_sync_batches:function(frm,cdt,cdn){
        frm.doc.items.forEach(item=>{
            if(item.custom_sub_batch){
                frappe.call({
                    method:"frappe.client.get_list",
                    args:{
                        doctype:"Batch",
                        order_by:"creation ASC",
                        fields:["name"],
                        filters:{"item":item.item_code,"custom_sub_batch":item.custom_sub_batch},
                        limit_page_length:1
                    },
                    callback:(r)=>{
                        if(r.message && r.message.length>0){
                            frappe.model.set_value(item.doctype, item.name,"batch_no",r.message[0].name)
                        }
                    }
                })
            }
        })
    }
})

File: tcb_manufacturing_customizations/doc_events/stock_reconciliation.py
import frappe
import time

def sync_sub_batch(doc, method=None):
    time.sleep(1)
    for item in doc.items:
        if item.use_serial_batch_fields and item.custom_sub_batch:
            latest_batch = frappe.db.get_all("Batch",filters={"item":item.item_code},order_by = "creation desc",fields = ["name"],limit=1)
            for batches in latest_batch:
                batch = frappe.get_doc("Batch",batches.name)
                sub_batch = batch.custom_sub_batch or None
                if not sub_batch:
                    frappe.db.set_value("Batch",batches.name,"custom_sub_batch",item.custom_sub_batch)
            
    frappe.db.commit()